### ISSUE TITLE
DNS Flag Values differ between draft-04 and RFC versions of Compactor

### ISSUE THREAD

**Issue by huitema** [dns-stats/compactor#62]
I am closely examining the output of the compactor for a test file, and comparing the latest version value (RFC) to the previous version (draft-04). I find some interesting differences:

| Old DNS Flag  | New DNS Flag|
| ------------ | -------------- |
| 22 (QRD+QZ+QAD) | 10 (QRA+QAD) |
| 7 (QZ+QAD+QCD)| 3 (QAD+QCD) |
| 31(QRD+QRA+QZ+QAD+QCD) | 15 (QRA+QZ+QAD+QCD) |

The differences appear systematic. I only dumped the first 16 transactions in the file, these were the only flag configurations present, and the old-new patterns were always the same.

Is there a simple explanation?

CORRECTION: Yes there is a simple explanation. I was looking at the SIG flag, not the DNS flag.

**Comment by huitema** [comment]
@johndickinson I am observing this pattern accross my entire test file. This is interesting, because my goal is to ensure that processing a file compressed using the draft-04 version of compressor gets the same results as processing the same file compressed with the RFC version. I used the "-n" parameter when compressing the file. Should I have used something else?

**Comment by huitema** [comment]
I tabulated the results. The "Old" column is the value of the qr-sig-flags parameter in entries of the QueryResponseSignature table parsed with the old (draft-04) compressor. The "New" column is the corresponding value for the same transaction in the table parsed with the new (RFC8618) compressor. The count column is the number of times this particular combination of old and new value of qr-sig-flags is observed in the whole file.


Old | New | Count
-- | -- | --
0x1f | 0x0f | 3700
0x07 | 0x03 | 1098
0x05 | 0x01 | 59
0x0d | 0x05 | 26
0x16 | 0x0a | 25
0x17 | 0x0b | 16
0x0f | 0x07 | 16
0x06 | 0x02 | 3
0x22 | 0x32 | 1
0x23 | 0x33 | 1

I wonder what causes that. Any idea?

**Comment by huitema** [comment]
Working on this again, it is most probably an issue in my code. The discrepancy seems to only happen on Windows.

**Comment by huitema** [comment]
OK, I get it now. The definitions of the "qr-sig-flags" change between draft-04 and RFC:

bit | draft | RFC
--- | ----- | ---
0   | 1 if a Query is present | 1 if a Query was present
1   | 1 if a Response is present |  1 if a Response was present
2   | 1 if one or more Question is present | 1 if a Query was present and it had an OPT RR
3   | 1 if a Query is present and it has an OPT Resource Record | 1 if a Response was present and it had an OPT RR
4   | 1 if a Response is present and it has an OPT Resource Record | 1 if a Query was present but had no Question
5   | 1 if a Response is present but has no Question | 1 if a Response is present but has no Question

What I saw was most probably the result of that change in spec, although there is still some strangeness. I verified that this completely explains the differences.
